### ISSUE TITLE
Rule Idea: Ensure yield on call to generator function

### ISSUE THREAD

**Issue by clark800** [eslint/eslint#5193]
> When will the rule warn?

When the code calls a generator function without a `yield` keyword. An optional setting may also make it warn when you yield the value of a function that is not a generator.

```
function * f (x) {
  return yield Promise.resolve(x * x);
}
console.log(f(2));
```

The output of the above script, when using "co", is "{}" rather than "4". After fixing the bug, the last line would be `console.log(yield f(2));`.

> Does the rule prevent an error or is stylistic?

This prevents errors that are particularly difficult to debug when using "co".

> Why the rule should be in the core instead of creating a custom rule.

It is useful for anyone who is using "co".

> Are you willing to create the rule yourself?

No

**Comment by eslintbot** [comment]
@clark800 Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Thanks for proposal. I have a feeling we already have something like that. @eslint/eslint-team am I making this up?

**Comment by platinumazure** [comment]
@ilyavolodin Something like [`require-yield`](http://eslint.org/docs/rules/require-yield)?

**Comment by ilyavolodin** [comment]
@mysticatea Does `require-yield` covers this proposal, and should it be updated?

**Comment by mysticatea** [comment]
No.

As my understanding, in this proposal, the rule warns call expressions if the callee is a generator function and without `yield`.
`require-yield` warns definitions of generator functions if the definition doesn't have `yield`.

Unfortunately, I don't think this proposal is feasible. ESLint is static analyzer, so it cannot know whether a callee is a generator function or not.

**Comment by clark800** [comment]
Generator functions are indicated with an asterisk in the function definition; there's no reason why that can't be statically determined. I understand that eslint works on each file individually, so it would not be able to apply this rule to "required" functions, but it should be possible for functions defined in the same file.

**Comment by ilyavolodin** [comment]
@clark800 Even withing the same file, there are problems linking calls to the functions. Here's a quick example:
```js
var a = function() {
};

var b = { test: a };

b["te" + "st"]();
```

Obviously, pretty much everything is possible with enough time and energy, but it's not easy.

**Comment by nzakas** [comment]
I don't think this is possible without a lot of false negatives. Taking just a simple example: if you have a class where one method is a generator, that means you now have to track all instances of that class to verify if the method is called. Another example: importing a generator from a module, you'd have no idea its a generator. 

Unfortunately, I don't see this as practical for inclusion in core, though you can always feel free to explore a custom rule on your own.
